### ISSUE TITLE
3.3.0: issue with generated setup.py in `entrypoints` module

### ISSUE THREAD

**Issue by kloczek** [takluyver/flit#437]
Because flit generats setup.py wgichhas line `from distutils.core import setup` instead `from setuptools import setup` generated setup.py is useless not only because it is not possible use setuptools<>sphinx integration but even with `distutils` that file is incorrect.
```console
+ /usr/bin/python3 setup.py build '--executable=/usr/bin/python3 -s'
/usr/lib64/python3.8/distutils/dist.py:274: UserWarning: Unknown distribution option: 'extras_require'
  warnings.warn(msg)
/usr/lib64/python3.8/distutils/dist.py:274: UserWarning: Unknown distribution option: 'python_requires'
  warnings.warn(msg)
running build
running build_py
creating build
creating build/lib
copying entrypoints.py -> build/lib
+ PBR_VERSION=0.3
+ SETUPTOOLS_SCM_PRETEND_VERSION=0.3
+ SPHINXOPTS=-j48
+ /usr/bin/python3 setup.py build_sphinx -b man --build-dir build/sphinx
/usr/lib64/python3.8/distutils/dist.py:274: UserWarning: Unknown distribution option: 'extras_require'
  warnings.warn(msg)
/usr/lib64/python3.8/distutils/dist.py:274: UserWarning: Unknown distribution option: 'python_requires'
  warnings.warn(msg)
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: invalid command 'build_sphinx'
```
Probably the same issue is with every other module which uses `flit`.

**Comment by kloczek** [comment]
Just FTT with quick patch:
```patch
--- a/setup.py~ 1970-01-01 01:00:00.000000000 +0100
+++ b/setup.py  2021-08-24 16:31:12.249073906 +0100
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
 # setup.py generated by flit for tools that don't yet use PEP 517

-from distutils.core import setup
+from setuptools import setup

 extras_require = \
 {":python_version == '2.7'": ['configparser>=3.5']}
```
above commands output is:
```console
+ /usr/bin/python3 setup.py build '--executable=/usr/bin/python3 -s'
running build
running build_py
creating build
creating build/lib
copying entrypoints.py -> build/lib
+ PBR_VERSION=0.3
+ SETUPTOOLS_SCM_PRETEND_VERSION=0.3
+ SPHINXOPTS=-j48
+ /usr/bin/python3 setup.py build_sphinx -b man --build-dir build/sphinx
running build_sphinx
Running Sphinx v4.1.2
making output directory... done
WARNING: html_static_path entry '_static' does not exist
loading intersphinx inventory from https://docs.python.org/objects.inv...
intersphinx inventory has moved: https://docs.python.org/objects.inv -> https://docs.python.org/3/objects.inv
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: [new config] 2 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
writing... entrypoints.1 { api } done
build succeeded, 1 warning.
```

**Comment by kloczek** [comment]
BTW `distutils` is obsoleted by https://www.python.org/dev/peps/pep-0632/

**Comment by bpabel** [comment]
It's been deprecated, but won't be removed until python 3.12, whose release is over two years away. Legacy non-pep517 setup.py builds (using `distutils` or `setuptools`) are going to go away along with `distutils` and legacy `setuptools` usage.

**Comment by kloczek** [comment]
`setuptools` is not growing anyway away as long as it provides crucial tooling infrastructure which none of new modules hooked under PEP517 are not able to replace, and as `stuptools` just Works™️.
'distutils' offer small subset of operations which `setuptools` can offer, it is not pluggable and it is used now by handful of modules. That subset recently slightly increased only because few modules started using `flit`.
In other words `flit` in current form somehow is trying to rollback some already done changes.
According to observation of my rpm packages `distutils` before `flit` started generating `distutils` depended setup.py files was used in ~1%. Since than this doubled to ~2%.
What is the sense of sticking with `distutils`? I have no idea ..

**Comment by takluyver** [comment]
The aim of using distutils there is to discourage people from relying on the generated `setup.py` - it was intended as a stopgap measure for older versions of pip, and it's going to go away.

https://github.com/takluyver/flit/blob/734ab12fc5bfadf7e37dc303d44fd517f7ebda1b/flit/sdist.py#L16-L18

**Comment by kloczek** [comment]
> The aim of using distutils there is to discourage people from relying on the generated `setup.py`

Why do you want to do that if `flit` cannot do the same things which `setuptools` do and all operations which `flit` provdes is possible to implemt using `setuptools`?

**Comment by bpabel** [comment]
You should read PEP517 and PEP518. It might help you understand the motivations to move towards a standardized build interface that doesn't defacto depend on setuptools.

**Comment by kloczek** [comment]
> You should read PEP517 and PEP518. It might help you understand the motivations to move towards a standardized build interface that doesn't defacto depend on setuptools.

I've read both documets and I was not able to find anything about moticvations (whatever in this context it mean in tis case).
Maybe it would be better if you could point please on exaxt parts of those documents?
What can solve on this case discoraging use setuptools is generally it works?
I can understand that there is a lot of legacy in many modules which are not using recent `setuptools` possibilities but discarding  whole `setuptoosl` in case when none of other tooling is able to do what `setuptools` is able to do is at least a bit odd.

**Comment by bpabel** [comment]
The problems with distutils/setuptools and legacy build systems is covered in the abstract and rationale sections of both peps.

**Comment by kloczek** [comment]
Again .. currently setuptools is able to hadne new approach.
Just look on setup.py, setup.cfg, pyproject.toml files for example in pip modile https://github.com/pypa/pip/
Please let me know if you see anything missing in that example ..

With current state og the `pip` there is no any issues with using setuptools<>sphinx integration.
If you want .hwl file you can generate it without `build` modules. Still it is possible to handle buiilding DSOs using for example `cythoon` (about which .wheel doesn';t know anything). All modules which are producing DSOs it is about 1/12 of all modules. `flit` and `poetry` do not provide any way to maintaind such modules.

What is IMO missing in PEP617 is standarication of metadata description. `flit` uses own [] sections and the same `poetry` instead reusing `setuptools` descriptions.
